Mathlib/Data/Vector/MapLemmas.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2023 Alex Keizer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Keizer
+-/
+import Mathlib.Data.Vector.Basic
+import Mathlib.Data.Vector.Snoc
+
+
+/-!
+  This file establishes a set of normalization lemmas for `map`/`mapAccumr` operations on vectors
+-/
+
+namespace Vector
+
+/-!
+## Fold nested `mapAccumr`s into one
+-/
+section Fold
+
+section Unary
+variable (xs : Vector α n) (f₁ : β → σ₁ → σ₁ × γ) (f₂ : α → σ₂ → σ₂ × β)
+
+@[simp]
+theorem mapAccumr_mapAccumr :
+    mapAccumr f₁ (mapAccumr f₂ xs s₂).snd s₁
+    = ((mapAccumr (fun x s =>
+        (((f₁ (f₂ x s.snd).snd s.fst).fst, (f₂ x s.snd).fst), (f₁ (f₂ x s.snd).snd s.fst).snd)
+      ) xs (s₁, s₂)).fst.fst, (mapAccumr (fun x s =>
+        (((f₁ (f₂ x s.snd).snd s.fst).fst, (f₂ x s.snd).fst), (f₁ (f₂ x s.snd).snd s.fst).snd)
+      ) xs (s₁, s₂)).snd) := by
+  induction xs using Vector.revInductionOn generalizing s₁ s₂ <;> simp_all
+
+@[simp]
+theorem mapAccumr_map (f₂ : α → β) :
+    (mapAccumr f₁ (map f₂ xs) s) = (mapAccumr (fun x s => f₁ (f₂ x) s) xs s) := by
+  induction xs using Vector.revInductionOn generalizing s <;> simp_all
+
+@[simp]
+theorem map_mapAccumr (f₁ : β → γ) :
+    (map f₁ (mapAccumr f₂ xs s).snd) = (mapAccumr (fun x s =>
+        ((f₂ x s).fst, f₁ (f₂ x s).snd)
+      ) xs s).snd := by
+  induction xs using Vector.revInductionOn generalizing s <;> simp_all
+
+@[simp]
+theorem map_map (f₁ : β → γ) (f₂ : α → β) :
+    map f₁ (map f₂ xs) = map (fun x => f₁ <| f₂ x) xs := by
+  induction xs using Vector.inductionOn <;> simp_all
+
+end Unary
+
+section Binary
+variable (xs : Vector α n) (ys : Vector β n)
+
+@[simp]
+theorem mapAccumr₂_mapAccumr_left (f₁ : γ → β → σ₁ → σ₁ × ζ) (f₂ : α → σ₂ → σ₂ × γ) :
+    (mapAccumr₂ f₁ (mapAccumr f₂ xs s₂).snd ys s₁)
+    = (
+        (mapAccumr₂ (fun x y s =>
+          (((f₁ (f₂ x s.snd).snd y s.fst).fst, (f₂ x s.snd).fst), (f₁ (f₂ x s.snd).snd y s.fst).snd)
+        ) xs ys (s₁, s₂)).fst.fst,
+        (mapAccumr₂ (fun x y s =>
+          (((f₁ (f₂ x s.snd).snd y s.fst).fst, (f₂ x s.snd).fst), (f₁ (f₂ x s.snd).snd y s.fst).snd)
+        ) xs ys (s₁, s₂)).snd) := by
+  induction xs, ys using Vector.revInductionOn₂ generalizing s₁ s₂ <;> simp_all
+
+@[simp]
+theorem map₂_map_left (f₁ : γ → β → ζ) (f₂ : α → γ) :
+    map₂ f₁ (map f₂ xs) ys = map₂ (fun x y => f₁ (f₂ x) y) xs ys := by
+  induction xs, ys using Vector.revInductionOn₂ <;> simp_all
+
+@[simp]
+theorem mapAccumr₂_mapAccumr_right (f₁ : α → γ → σ₁ → σ₁ × ζ) (f₂ : β → σ₂ → σ₂ × γ) :
```

Comment:
I think I might have changed my mind about the use of lets. It's very unreadable right now. Maybe the lets should be reintroduced on second thoughts.